src/main/scala/scala/xml/parsing/MarkupParserCommon.scala

```diff
@@ -247,7 +247,7 @@ private[scala] trait MarkupParserCommon extends TokenTests {
       while (true) {
         if (ch == head && peek(rest))
           return handler(positioner(), sb.toString)
-        else if (ch == SU)
+        else if (ch == SU || eof)
           truncatedError("") // throws TruncatedXMLControl in compiler
 
```

Comment:
It doesn't seem like adding the `eof` here triggers an OOM for any of your tests.  I couldn't find an example that triggered the OOM for an XML processing instruction (PI) or a CDATA section -- they're the only ones that use `xTakeUntil` .  I guess it's ok to cargo-cult an `eof` check everywhere, but it seems you were trying to avoid that.